pkg/util/aad/aad.go

```diff
@@ -95,3 +99,44 @@ func GetToken(ctx context.Context, log *logrus.Entry, oc *api.OpenShiftCluster,
 
 	return token, nil
 }
+
+// GetClusterSPObjectID returns a service principal objectID for a cluster.
+// It retries in case of Authorization_IdentityNotFound error as it is likely due to AAD sync delay.
+func GetClusterSPObjectID(ctx context.Context, log *logrus.Entry, oc *api.OpenShiftCluster, token *adal.ServicePrincipalToken) (string, error) {
+	spp := &oc.Properties.ServicePrincipalProfile
+
+	spGraphAuthorizer := autorest.NewBearerAuthorizer(token)
+	applications := graphrbac.NewApplicationsClient(spp.TenantID, spGraphAuthorizer)
+
+	timeoutCtx, cancel := context.WithTimeout(ctx, 15*time.Minute)
+	defer cancel()
+
+	var clusterSPObjectID string
+	var err error
+	// NOTE: Do not override err with the error returned by wait.PollImmediateUntil.
+	// Doing this will not propagate the latest error to the user in case when wait exceeds the timeout
+	wait.PollImmediateUntil(10*time.Second, func() (bool, error) {
+		res, innerErr := applications.GetServicePrincipalsIDByAppID(ctx, spp.ClientID)
+		if innerErr != nil {
+			isAuthorizationIdentityNotFound := strings.Contains(innerErr.Error(), "Authorization_IdentityNotFound")
+
+			// populate err with a user-facing error that will be visible if we're not successful.
+			log.Info(innerErr)
+			err = api.NewCloudError(http.StatusBadRequest, api.CloudErrorCodeInvalidServicePrincipalCredentials, "properties.servicePrincipalProfile", "The provided service principal credentials are invalid.")
```

Comment:
Any suggestions for better user-facing error message? It's currently the same as from `GetToken`

Reply:
Or should we just return plain error from `GetServicePrincipalsIDByAppID` (will resutl in "Internal server error") as we did before my changes?

Reply:
No need to add user-facing error here as there is nothing user can do